contracts/token/ERC6909/draft-ERC6909.sol

```diff
@@ -0,0 +1,123 @@
+    }
+
+    function balanceOf(address owner, uint256 id) public view virtual override returns (uint256) {
+        return _balances[id][owner];
+    }
+
+    function allowance(address owner, address spender, uint256 id) public view virtual override returns (uint256) {
+        return _allowances[owner][spender][id];
+    }
+
+    function isOperator(address owner, address spender) public view virtual override returns (bool) {
+        return _operatorApprovals[owner][spender];
+    }
+
+    function approve(address spender, uint256 id, uint256 amount) external virtual override returns (bool) {
+        address caller = _msgSender();
+        _allowances[caller][spender][id] = amount;
+
+        emit Approval(caller, spender, id, amount);
+        return true;
+    }
+
+    function setOperator(address spender, bool approved) external virtual override returns (bool) {
+        address caller = _msgSender();
+        _operatorApprovals[caller][spender] = approved;
+
+        emit OperatorSet(caller, spender, approved);
+        return true;
+    }
+
+    function transfer(address receiver, uint256 id, uint256 amount) external virtual override returns (bool) {
+        _update(_msgSender(), receiver, id, amount);
+        return true;
+    }
+
+    function transferFrom(
+        address sender,
+        address receiver,
+        uint256 id,
+        uint256 amount
+    ) external virtual override returns (bool) {
+        address caller = _msgSender();
+        if (caller != sender && !isOperator(sender, caller)) {
+            uint256 currentAllowance = _allowances[sender][caller][id];
+            if (currentAllowance != type(uint256).max) {
+                if (currentAllowance < amount) {
+                    revert ERC6909InsufficientAllowance(caller, currentAllowance, amount, id);
+                }
+                unchecked {
+                    _allowances[sender][_msgSender()][id] = currentAllowance - amount;
+                }
+            }
+        }
+
+        _transfer(sender, receiver, id, amount);
+        return true;
+    }
+
+    function _transfer(address from, address to, uint256 id, uint256 amount) internal {
+        if (from == address(0)) {
+            revert ERC6909InvalidSender(address(0));
+        }
+        if (to == address(0)) {
+            revert ERC6909InvalidReceiver(address(0));
+        }
+        _update(from, to, id, amount);
+    }
+
+    function _update(address from, address to, uint256 id, uint256 amount) internal virtual {
+        address caller = _msgSender();
+
+        if (from != address(0)) {
+            uint256 fromBalance = _balances[id][from];
+            if (fromBalance < amount) {
+                revert ERC6909InsufficientBalance(from, fromBalance, amount, id);
+            }
+            unchecked {
+                _balances[id][from] -= amount;
+            }
+        }
+        if (to != address(0)) {
+            _balances[id][to] += amount;
```

Comment:
Just realized this can't be `unchecked` given it's not bound by a `totalSupply` as ERC20 does, which means that multiple accounts can hold more than the maximum uint256 value.

I wonder if this is an issue we should worry about. An example where it can break is for off-chain indexers who may want to track the total supply but are restricted to an uint256 (as the graph)

Reply:
ERC1155 is the same by default. Given that the ERC doesn't restrict this at all, and that it is already done in ERC1155, I think it is fine.